### ISSUE TITLE
Ledger Signing: Upgrade from ethereumjs-tx to @ethereumjs/tx not working

### ISSUE THREAD

**Issue by wspi** [ethereumjs/ethereumjs-monorepo#1228]
I have a code that signs a transaction using ethereumjs-tx and it works, then I've decided to migrate to @ethereumjs/tx, the code is almost equal, but when I try to send a transaction sent using @ethereumjs/tx I get the error `Error: Returned error: insufficient funds for gas * price + value`

I'm using a HSM to sign the transactions, so I set the r, s, v manually

here are some parts of both codes

Working:
```
const eth_tx = require('ethereumjs-tx');

sig = await sign(AddrHash);

txParams = {
    nonce: await web3.eth.getTransactionCount(ethAddr),
    gasPrice: '0x0918400000',
    gasLimit: 160000,
    to: addr,
    value: web3.utils.toHex(web3.utils.toWei('0.0001', "ether")),
    data: '0x00',
    r: sig.r.toBuffer(),
    s: sig.s.toBuffer(),
    v: v
}

tx = new eth_tx.Transaction(txParams, { chain: 'rinkeby' });

let txHash = tx.hash(false);
sig = await sign(txHash);

tx.r = sig.r.toBuffer();
tx.s = sig.s.toBuffer();
tx.v = new bn(v).toBuffer();

const serializedTx = tx.serialize().toString('hex');
web3.eth.sendSignedTransaction('0x' + serializedTx)
.on('confirmation', (confirmationNumber, receipt) => {})
.on('receipt', (txReceipt) => {
    return txReceipt.transactionHash;
})
.on('error', error => console.log(error));
```

not working:
```
const eth_tx = require('@ethereumjs/tx');
const Common = require('@ethereumjs/common').default;


const common = new Common({ chain: 'rinkeby', hardfork: 'berlin'});

sig = await sign(AddrHash);

txParams = {
    nonce: await web3.eth.getTransactionCount(ethAddr),
    gasPrice: '0x0918400000',
    gasLimit: 160000,
    to: addr,
    value: web3.utils.toHex(web3.utils.toWei('0.0001', "ether")),
    data: '0x00',
    r: sig.r.toBuffer(),
    s: sig.s.toBuffer(),
    v: v
}

tx = eth_tx.Transaction.fromTxData(txParams, { common });

let txHash = tx.hash(false);
sig = await sign(txHash);

txParamsSigned = {
  nonce: tx.nonce,
  gasPrice: tx.gasPrice,
  gasLimit: tx.gasLimit,
  to: tx.to,
  value: tx.value,
  data: tx.data,
  r: sig.r.toBuffer(),
  s: sig.s.toBuffer(),
  v: new bn(v).toBuffer()
};

signed_tx = eth_tx.Transaction.fromTxData(txParamsSigned, { common });
const serializedTx = signed_tx.serialize().toString('hex');

web3.eth.sendSignedTransaction('0x' + serializedTx)
.on('confirmation', (confirmationNumber, receipt) => {})
.on('receipt', (txReceipt) => {
    return txReceipt.transactionHash;
})
.on('error', error => console.log(error));
```

any ideas?

**Comment by jochem-brouwer** [comment]
My hunch is that your hardware wallet is not receiving the right ChainID of rinkeby. Could you try to `getSenderAddress()` on the transaction and check if this reports the expected address?

**Comment by wspi** [comment]
@jochem-brouwer getSenderAddress() is returning the wrong address

have I not set the `Common` properly?

**Comment by jochem-brouwer** [comment]
I think you are actually signing the wrong message.

```javascript
const eth_tx = require('@ethereumjs/tx');
const Common = require('@ethereumjs/common').default;


const common = new Common({ chain: 'rinkeby', hardfork: 'berlin'});

txParams = {
    nonce: await web3.eth.getTransactionCount(ethAddr),
    gasPrice: '0x0918400000',
    gasLimit: 160000,
    to: addr,
    value: web3.utils.toHex(web3.utils.toWei('0.0001', "ether")),
    data: '0x00'
}

tx = eth_tx.Transaction.fromTxData(txParams, { common });

let messageToSign = tx.getMessageToSign(false); // Might also want to try this with `true`
sig = await sign(txHash);

txParamsSigned = {
  nonce: tx.nonce,
  gasPrice: tx.gasPrice,
  gasLimit: tx.gasLimit,
  to: tx.to,
  value: tx.value,
  data: tx.data,
  r: sig.r.toBuffer(),
  s: sig.s.toBuffer(),
  v: new bn(v).toBuffer()
};

signed_tx = eth_tx.Transaction.fromTxData(txParamsSigned, { common });
const serializedTx = signed_tx.serialize().toString('hex');

web3.eth.sendSignedTransaction('0x' + serializedTx)
.on('confirmation', (confirmationNumber, receipt) => {})
.on('receipt', (txReceipt) => {
    return txReceipt.transactionHash;
})
.on('error', error => console.log(error));
```

What happens if you try this?

**Comment by wspi** [comment]
same thing, address does not match, and `Error: Returned error: insufficient funds for gas * price + value` error

both with tx.getMessageToSign(false); and true

**Comment by wspi** [comment]
I tested removing the r, s, v from the txParams but then I get the error `Error: This transaction is not signed`

**Comment by jochem-brouwer** [comment]
Do you get a different address now? How does the `sign` method look? 

You might want to join our [discord](https://discord.gg/xPzDeb3REy) so we can have a quick chat about this.

**Comment by bkbooth** [comment]
@wspi or @jochem-brouwer, was there a solution to this? I'm seeing this same error with transactions that I'm trying to sign on a Ledger.

My `common` appears to be configured correctly (using 'ropsten'), but when I send the transaction to be signed on the Ledger it's showing 'mainnet'. After signing the transaction and re-building the tx using the `v`, `r` and `s`, it shows as signed and verified. When I send it I get the same error as the OP.

**Comment by wspi** [comment]
@bkbooth nope, we haven't been able to sign correctly the TX, I've noticed that the hash function on the monorepo version does not have the same functionality as in the legacy lib, also getMessageToSign is similar, but not the same, I believe that may be the cause of not been able to sign it correctly, tried some things like coding the same logic as the legacy lib, but wasn't able to broadcast the tx

**Comment by holgerd77** [comment]
@wspi Ok, I did the counter-test here and send (more or less) the same tx from the example to Goerli (don't have Ropsten ETH atm) and tried to reproduce the setup as close as possible, but then I used the signing functionality from within the `Transaction` library with `const signedTx = tx.sign(PRIV_KEY)`.

This actually worked and the following tx was sent out: https://goerli.etherscan.io/tx/0x653da6acdf6f16f0cd4a8d6e1a630d5dd2665318094fc9d8609c6286900c975a

Hope this helps to narrow things in a bit.

What you can actually test is to get the signature from your HSM and generate a signature with the tx library as I did above and then compare the `v`, `r` and `s` values, these should be exposed by the `signedTx` object. Are these identical to the onces received from the HSM?

**Comment by holgerd77** [comment]
Also an idea for testing: can you try to create a *new* transaction object with your `v`, `r`, `s` values and not just assign with `tx.v=...` and so on?

**Comment by holgerd77** [comment]
Ok, I did a test installation of the old `ethereumjs-tx` library and compared the unsigned tx hash outputs. For my example (taken from the above) the resulting hash strings are the same:

@ethereumjs/tx, tx.getMessageToSign(): eba71536e6b07ffd5fcf6faaada215a476d03e84dc5ca303f598b2d17c826c5c
ethereumjs-tx, tx.hash(false):         eba71536e6b07ffd5fcf6faaada215a476d03e84dc5ca303f598b2d17c826c5c

So guys, you need to really help us here with some additional testing on your own and posting in-between results so that we can get this solved. What are the r,s,v values you are getting for both library versions? Or the serialized tx string? Are these identical?

**Comment by holgerd77** [comment]
Hi guys, we've identified the problem - see #1278 for context - as a hotfix you can use the following to make this work again:

```typescript
import { BN } from 'ethereumjs-util'
const tx = Transaction.fromTxData(txParams, { common, freeze: false }) // Allow the tx to be modified
tx.v = new BN(4) // Rinkeby chain ID
tx.serialize()
```

**Comment by wspi** [comment]
hi @holgerd77 actually I believe it's not the issue I'm having, probably it may solve @bkbooth issue, but mine wasn't solved by this



> Ok, I did a test installation of the old `ethereumjs-tx` library and compared the unsigned tx hash outputs. For my example (taken from the above) the resulting hash strings are the same:
> 
> @ethereumjs/tx, tx.getMessageToSign(): eba71536e6b07ffd5fcf6faaada215a476d03e84dc5ca303f598b2d17c826c5c
> ethereumjs-tx, tx.hash(false): eba71536e6b07ffd5fcf6faaada215a476d03e84dc5ca303f598b2d17c826c5c
> 
> So guys, you need to really help us here with some additional testing on your own and posting in-between results so that we can get this solved. What are the r,s,v values you are getting for both library versions? Or the serialized tx string? Are these identical?

the issue here is that tx.hash(false) ignores the signature data (v, s, r) while tx.getMessageToSign() does not, so I get different hashs as my tx already has the v, s, r set

and if I create the tx without the v, s, r the signing does not work, even on the ethereumjs-tx

I've tried creating the tx without the v, s, r and hashing before setting the v, s, r but it didn't work also

**Comment by wspi** [comment]
@holgerd77 just giving more info, what happens to me is that the senderAddress before signing the tx is mine, but after signing the same way I do on the `ethereumjs-tx` the senderAddress changes, and does not belong to me, so when trying to broadcast the TX I get this error `insufficient funds for gas * price + value`

**Comment by bkbooth** [comment]
@holgerd77 that works for me, thanks! I actually went to try that when I was stuck on this problem before reverting back to `ethereumjs-tx`, but I'm using TypeScript, and TS is complaining that `v` and other properties on `tx` are `readonly`. Looks like the types need to be changed for when the tx is created with `{ ..., freeze: false }`.

I noticed in #1278 that you suggested internally setting `v` using `common.chainId` for unsigned legacy transaction. I thought I'd weigh in and mention that was the behaviour I was _expecting_.

@wspi just checking that you tried the solution @holgerd77 suggested? I was seeing the same problem with the wrong sender address after signing (as well as the chain ID showing up wrong on my Ledger before signing). After explicitly setting `tx.v`, the chain ID is correct on my Ledger, and the sender address is correct after signing.

**Comment by holgerd77** [comment]
@bkbooth yeah, somewhat complicated topic, I guess one main problem is EIP-155 staying somewhat blurry if this defines an unsigned tx by itself (and - as @alcuadrado pointed out - this is likely not the case even if often interpreted in this way) *or* just what data is used for hashing respectively for signing, also see the follow-up comment https://github.com/ethereumjs/ethereumjs-monorepo/issues/1278#issuecomment-853653940 I made on the thread.

Phew. But I was drawn back and forth myself.

Anyhow. We have found a better - somewhat official solution (see also the comment from above for some explanation), will repost here. This should avoid any TypeScript issues.

```typescript
import { rlp } from 'ethereumjs-util'
import Common from '@ethereumjs/common'
import { Transaction } from '@ethereumjs/tx'

const common = new Common({ chain: 'goerli', hardfork: 'berlin'})
const tx = Transaction.fromTxData({}, { common })

const message = tx.getMessageToSign(false)
const serializedMessage = rlp.encode(message) // use this for the ledger input
```

Re-stating here. The API on `tx.getMessageToSign()` is suboptimal right now by not doing the RLP serialization step inside. We can't switch this though imminently right now and need to wait until the next breaking release, which is some months away. For now I will therefore address by doing better documentation on this in the code docs and eventually other places.

**Comment by holgerd77** [comment]
@wspi you need to take an unsigned tx without having `v`, `r`, `s` set for signing, trying to sign a signed tx (with `v`, `r`, `s` set, so this means: signed) won't work.

**Comment by bkbooth** [comment]
@holgerd77 that suggestion works, much nicer code (& TypeScript happy), thanks 🙏

**Comment by wspi** [comment]
@holgerd77 @bkbooth I've tried both ways, using the rlp.encode(message) é got `ValidationException: Digest is invalid length for algorithm ECDSA_SHA_256.` when signing

**Comment by holgerd77** [comment]
Solved by #1283 will close.
